### DISCUSSION TITLE
Custom hook and access to WidgetRef

### DISCUSSION THREAD

**Discussion by dastein1** [rrousselGit/flutter_hooks#324] [General]
Say that I have a custom hook in which I want to interact with riveropd providers amongst other things. 
Can I just pass in the WidgetRef as a param to the hook then?

    useMyCustomHook(WidgetRef ref) {
      final data = ref.watch(someProvider);

      useEffect(() {
        // do something on data change and maybe call some other provider
        // ref.read(someOtherProvider.notifier).callSomeAction();
      }, [data, ref]);

      // do more stuff 
    }

This satisfies all rules for hooks and seem to work just fine in my testing.
But I'm interested if this is actually considered an anti-pattern that should be avoided.

**Comment by davidmartos96** [comment]
Passing the ref is the recommended solution, there is nothing wrong with it.

**Comment by davidmartos96** [comment]
I don't think it's needed as a dependency in the useEffect, but I might be wrong

**Reply by dastein1** [reply]
@davidmartos96 Thanks. In my real world example I'm accessing _data_ in the effect and so need it as a dependency to rerun as it changes. Not sure if it is safe to omit _ref_. Thoughts?

**Reply by davidmartos96** [reply]
When would you like your useEffect to be called again?

**Reply by rrousselGit** [reply]
Not having ref as dependency should be fine.

**Reply by rrousselGit** [reply]
But you probably shouldn't use ref inside effects

Try using ref outside of the effect. Then the values you obtained should be passed as dependency to the effect

**Reply by dastein1** [reply]
@rrousselGit Thanks. In that example I am watching 1 provider outside and passing that data into the effect. In that effect however I need to call a statenotifier's method and therefore used _ref.read()_. That should be legit, isn't it?
